portfolio/src/main/java/com/google/sps/servlets/GetCommentsServlet.java

```diff
@@ -36,27 +36,22 @@
 public class GetCommentsServlet extends HttpServlet {
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    // Get the number of comments requested.
-    String numCommentsString = request.getParameter("num");
+    String commentOrderString = request.getParameter("order");
 
-    // Convert the input to an int.
-    int numComments;
-    try {
-      numComments = Integer.parseInt(numCommentsString);
-    } catch (NumberFormatException e) {
-      System.err.println("Could not convert to int: " + numCommentsString);
-      return;
+    Query query;
+    if (commentOrderString.equals("oldest")) {
+      query = new Query("Comment").addSort("timestamp", SortDirection.ASCENDING);
+    } else {
+      query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
     }
-    Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
 
-    // Query Datastore for with limit for number of comments requested
+    // Query Datastore for all comments (to support pagination)
```

Comment:
Nice job on taking the initiative to add pagination! And apologies in advance for the super-long feedback comment here.

If I'm reading your code right, what you're basically doing is:
* Always returning ALL comments on the backend
* Loading ALL comments on the frontend, and displaying only a portion, based on state variables you maintain

As you can imagine, the more comments you have, the slower the page will be.

One very quick fix is to only load the comments once, and then have the pagination buttons just use the data you've already loaded... then you can provide a refresh button (and refresh any time a comment is added), and you're set!

... that kind of thing would be sufficient if you knew you'd only ever get like 100 comments total, and limited it to 100 on the backend side. If what I'm about to describe below is not doable, or too much work, then just set a max number of comments to load and try not to reload it with every click on the page.

.... Now... Imagine you have millions, or hundreds of millions of comments. Popular youtube videos, as an example, have high hundreds of thousands to millions of comments. So, it goes without saying that loading all of them would be intractable.

## A General Explanation (actual thing you can/should do linked below):

The more standard way that pagination is solved is by making the backend API support pagination. A request into a backend API that supports pagination might look like:

```
REQUEST:
{
    // ... all the fields needed for the request, like if it's comments
    // on a post, there'd be a postId. This is not relevant here...

    // Pagination-specific stuff.
    pageSize: 50,
    pageToken: <discussed below>
}
```

The response would look like what's below. It's basically what you have now (a list of comments), but wrapped into a response object that includes some pagination specifics.

```
RESPONSE:
{
    // What you already have.
    comments: [ <comment1>, <comment2>, ...]

    // Pagination stuff.
    nextPageToken: "abcd1234...whatever"
}
```

When you first call the API when your page loads, you would just pass it a page size, and leave `pageToken` blank. Then, when it's time to load the next page, you would call the API again, except this time, you would pass in the `nextPageToken` as a `pageToken`, so it knows where to start. If there is no `nextPageToken` returned, you know there's no more pages.

In the backend, you need to make it so it doesn't load every comment. In the backend, you would need it to load only the tokens that belong on the next page.

For this to work, your data store needs to support indexing (say by timestamp). So, let's say your `pageToken` is "timestamp + comment ID". Then, you could query for comments greater than or equal the timestamp, and start "counting" only when you found the comment ID in question.

Then, you'd load up one more comment than your page size (so pageSize = 50 means you load up 51 total), but instead of returning that 51st comment, you convert it into a nextPageToken (grab its timestamp and ID and send it as the next page token). Then, when the user clicks, the next request will know where to begin.

There are a bunch of nuances I'm glazing over, like what happens when comments are deleted or things change. This is too complex to address in an overly-long code review comment already. :-)

The downside of this approach is that implementing "prev" is quite difficult... though the good news is, when dealing with unbounded things like comments, you don't usually need it, and you end up with something like facebook or youtube, where you just keep loading more. There ARE ways to do "prev" with this approach, but they're a bit more painful.

## What AppEngine datastore recommends you do

Turns out AppEngine's DataStore you're using has this functionality built in already:

https://cloud.google.com/datastore/docs/concepts/queries#cursors_limits_and_offsets

This includes some example code for achieving the result. It looks like it makes you return its internal cursor directly to the frontend, which has its own intricacies, but I believe you should be able to make this work with relative ease.

Reply:
Super fun for me to try and figure this out today. I actually really missed the Prev button so decided to try and implement it. I think I was able to do it with the array method I mentioned in the chat.